File: src/main/java/com/mpe85/grampa/rule/Rule.java
package com.mpe85.grampa.rule;

import java.util.List;

import com.mpe85.grampa.visitor.RuleVisitor;

/**
 * Defines a parser rule.
 * 
 * @author mpe85
 *
 * @param <T>
 *            the type of the stack elements
 */
public interface Rule<T> {
	
	/**
	 * Gets the child rules of the rule.
	 * 
	 * @return a list of parser rules
	 */
	List<Rule<T>> getChildren();
	
	/**
	 * Gets the (first) child rule of the rule.
	 * 
	 * @return a parser rule
	 */
	Rule<T> getChild();
	
	/**
	 * Replaces a reference rule inside the list of child rules.
	 * 
	 * @param index
	 *            the position if the reference rule inside the list of child rules
	 * @param replacementRule
	 *            a replacement rule
	 * @return the replaced reference rule
	 */
	Rule<T> replaceReferenceRule(int index, Rule<T> replacementRule);
	
	/**
	 * Try to match the rule
	 * 
	 * @param context
	 *            a rule context.
	 * @return true if the rule matched successfully, otherwise false
	 */
	boolean match(RuleContext<T> context);
	
	/**
	 * Checks if the rule is part of a predicate rule.
	 * 
	 * @return true if the rule is a direct or indirect child of a predicate rule, otherwise false
	 */
	boolean isPredicate();
	
	/**
	 * Accepts a rule visitor.
	 * 
	 * @param visitor
	 *            a rule visitor
	 */
	void accept(RuleVisitor<T> visitor);
	
}

File: src/main/kotlin/com/mpe85/grampa/rule/Rule.kt
package com.mpe85.grampa.rule

import com.mpe85.grampa.context.ParserContext
import com.mpe85.grampa.visitor.RuleVisitor

/**
 * The parser rule interface.
 *
 * @author mpe85
 * @param[T] The type of the stack elements
 * @property[children] The child rules of the rule
 * @property[child] The (first) child rule of the rule
 * @property[testRule] Indicates if the rule is directly or indirectly part of a test rule.
 */
interface Rule<T> {

    val children: List<Rule<T>>
    val child: Rule<T>?
    val testRule: Boolean

    /**
     * Replace a reference rule inside the list of child rules.
     *
     * @param[index] The position if the reference rule inside the list of child rules
     * @param[replacementRule] A replacement rule
     * @return The replaced reference rule
     */
    fun replaceReferenceRule(index: Int, replacementRule: Rule<T>): Rule<T>

    /**
     * Try to match the rule
     *
     * @param[context] A rule context
     * @return true if the rule matched successfully
     */
    fun match(context: ParserContext<T>): Boolean

    /**
     * Accept a [RuleVisitor].
     *
     * @param visitor A rule visitor
     */
    fun accept(visitor: RuleVisitor<T>)
}
